src/gateway/handlers/handle-error.js

```diff
@@ -0,0 +1,89 @@
+// @flow
+import {extractError} from "../../shared/index.js";
+import {getLogger} from "../../ka-shared/index.js";
+import type {AmbiguousError} from "../../shared/index.js";
+import type {Request, Response, CustomErrorHandlerFn} from "../types.js";
+
+/**
+ * Handle an ambiguous error and determine an appropriate response.
+ *
+ * @param {string} overallProblem A simple description of what problem the error
+ * explains.
+ * @param {?CustomErrorHandlerFn} errorHandler A possible custom error handler
+ * that can generate a response for the error.
+ * @param {Request} req The request that was being handled when the error
+ * occurred.
+ * @param {Response} res The response that is being generated and to which the
+ * error is reported.
+ * @param {AmbiguousError} error The error that is to be handled.
+ */
+export const handleError = (
+    overallProblem: string,
+    errorHandler: ?CustomErrorHandlerFn,
+    req: Request,
+    res: Response,
+    error: AmbiguousError,
+): void => {
+    const logger = getLogger(req);
+    /**
+     * Something went wrong. Let's report it!
+     */
+    const simplifiedError = extractError(error);
+
+    /**
+     * We're definitely returning an error for this one.
+     */
+    res.status(500);
+
+    /**
+     * The calling code should handle if the original request was valid or
+     * not, so we just appease flow here.
+     */
+    const requestURL = typeof req.query.url === "string" ? req.query.url : "";
+
+    /**
+     * Before we return the basic 500 error, let's give our configuration
+     * chance to make a nicer error page.
+     */
+    try {
+        const overriddenResponse = errorHandler?.(
```

Comment:
How do we expect that this function to set response headers - by modifying the headers object? It seems like we should set the `Content-type: text/html` header by default, at least, so that the handler doesn't have to do it. But I do think that the error response handler is going to want to have the same functionality as the main render function - we want to be able to compute the vary headers and set the right Surrogate-Control headers so that Fastly caches the response correctly, otherwise users will just keep making failed requests again-and-again.

Reply:
Do we set headers on 500 errors like that?

Reply:
Yes, for errors handled internally we'll set those headers (same as we would for 404, 403, or other error pages) - if we have the opportunity to generate an error page it feels like we'd want to try and do it there, too!

Reply:
OK, makes sense.

Reply:
So, as I understand it, we have three main ways to error:

1. Render causes error, but webapp code catches and provides a nice 500 result with headers; this code never even knows; it just sends back the render result.
2. Render fails and we need to provide something half decent so we call the custom handler and the render-gateway service builds a nice result from that specific to webapp - Does this add minimal headers to ensure result is cached, assuming that it's going to happen every time for the same request? - what if it's a one off? Do we want to cache the result? I'm not sure - this bit us with those 502 errors as we cached nothing and then never SSR'd when it was only a temporal issue; that said, seems reasonable we still might want to.
- I ask because this is not going to use webapp frontend code; it will all be in the service, so it's not intended to be super complex. 
3. The custom handling fails and we just have to return something. At this point, we're likely not adding any headers since that's true last ditch and this doesn't know anything about specifics.